mrp_lot_number_propagation/models/mrp_bom.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2022 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+from odoo import _, api, fields, models, tools
+from odoo.exceptions import ValidationError
+
+
+class MrpBom(models.Model):
+    _inherit = "mrp.bom"
+
+    lot_number_propagation = fields.Boolean(
+        default=False,
+        help=(
+            "Allow to propagate the lot/serial number "
+            "from a component to the finished product."
+        ),
+    )
+    display_lot_number_propagation = fields.Boolean(
+        compute="_compute_display_lot_number_propagation"
+    )
+
+    @api.depends(
+        "type",
+        "product_tmpl_id.tracking",
+        "product_qty",
+        "product_uom_id",
+        "bom_line_ids.product_id.tracking",
+        "bom_line_ids.product_qty",
+        "bom_line_ids.product_uom_id",
+    )
+    def _compute_display_lot_number_propagation(self):
+        """Check if a lot number can be propagated.
+
+        A lot number can be propagated from a component to the finished product if:
+        - the type of the BoM is normal (Manufacture this product)
+        - the finished product is tracked by serial number
+        - the quantity of the finished product is 1 and its UoM is unit
+        - there is at least one bom line, with a component tracked by serial,
+          having a quantity of 1 and its UoM is unit
+        """
+        uom_unit = self.env.ref("uom.product_uom_unit")
+        for bom in self:
+            bom.display_lot_number_propagation = (
+                bom.type in self._get_lot_number_propagation_bom_types()
+                and bom.product_tmpl_id.tracking == "serial"
+                and tools.float_compare(
+                    bom.product_qty, 1, precision_rounding=bom.product_uom_id.rounding
+                )
+                == 0
+                and bom.product_uom_id == uom_unit
+                and bom._has_tracked_product_to_propagate()
+            )
+
+    def _get_lot_number_propagation_bom_types(self):
+        return ["normal"]
+
+    def _has_tracked_product_to_propagate(self):
+        self.ensure_one()
+        uom_unit = self.env.ref("uom.product_uom_unit")
+        for line in self.bom_line_ids:
+            if (
+                line.product_id.tracking == "serial"
+                and tools.float_compare(
+                    line.product_qty, 1, precision_rounding=line.product_uom_id.rounding
```

Comment:
I have an error here when testing on runboat:
```


Odoo Server Error

Traceback (most recent call last):
  File "/opt/odoo/odoo/addons/base/models/ir_http.py", line 237, in _dispatch
    result = request.dispatch()
  File "/opt/odoo/odoo/http.py", line 696, in dispatch
    result = self._call_function(**self.params)
  File "/opt/odoo/odoo/http.py", line 370, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo/odoo/http.py", line 358, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo/odoo/http.py", line 919, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo/odoo/http.py", line 544, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo/addons/web/controllers/main.py", line 1370, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/opt/odoo/addons/web/controllers/main.py", line 1362, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 406, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 391, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/opt/odoo/odoo/models.py", line 6266, in onchange
    snapshot0 = Snapshot(record, nametree, fetch=(not first_call))
  File "/opt/odoo/odoo/models.py", line 6110, in __init__
    self.fetch(name)
  File "/opt/odoo/odoo/models.py", line 6120, in fetch
    self[name] = record[name]
  File "/opt/odoo/odoo/models.py", line 5727, in __getitem__
    return self._fields[key].__get__(self, self.env.registry[self._name])
  File "/opt/odoo/odoo/fields.py", line 1043, in __get__
    self.compute_value(recs)
  File "/opt/odoo/odoo/fields.py", line 1199, in compute_value
    records._compute_field_value(self)
  File "/opt/odoo/odoo/models.py", line 4089, in _compute_field_value
    odoo.fields.determine(field.compute, self)
  File "/opt/odoo/odoo/fields.py", line 83, in determine
    return needle(*args)
  File "/mnt/data/odoo-addons-dir/mrp_lot_number_propagation/models/mrp_bom_line.py", line 25, in _compute_display_propagate_lot_number
    line.bom_id.display_lot_number_propagation
  File "/opt/odoo/odoo/fields.py", line 1043, in __get__
    self.compute_value(recs)
  File "/opt/odoo/odoo/fields.py", line 1199, in compute_value
    records._compute_field_value(self)
  File "/opt/odoo/addons/mail/models/mail_thread.py", line 410, in _compute_field_value
    return super()._compute_field_value(field)
  File "/opt/odoo/odoo/models.py", line 4089, in _compute_field_value
    odoo.fields.determine(field.compute, self)
  File "/opt/odoo/odoo/fields.py", line 83, in determine
    return needle(*args)
  File "/mnt/data/odoo-addons-dir/mrp_lot_number_propagation/models/mrp_bom.py", line 51, in _compute_display_lot_number_propagation
    and bom._has_tracked_product_to_propagate()
  File "/mnt/data/odoo-addons-dir/mrp_lot_number_propagation/models/mrp_bom.py", line 63, in _has_tracked_product_to_propagate
    and tools.float_compare(
  File "/opt/odoo/odoo/tools/float_utils.py", line 155, in float_compare
    rounding_factor = _float_check_precision(precision_digits=precision_digits,
  File "/opt/odoo/odoo/tools/float_utils.py", line 29, in _float_check_precision
    assert precision_rounding is None or precision_rounding > 0,\
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/odoo/odoo/http.py", line 652, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/odoo/odoo/http.py", line 317, in _handle_exception
    raise exception.with_traceback(None) from new_cause
AssertionError: precision_rounding must be positive, got 0.0
```
The rounding is set on the UOM, the UOM is set on the product, but it seems not to be set on the BOM line before the computation.

Reply:
I tested on the customer instance without error when opening the PR.
I am not sure how to reproduce this one, but I just pushed a fixup that could fix the problem.
If the error is still present it would be good if we can test together.
Thanks for the review.

Reply:
As discussed, this was a bug in `mrp_bom_attribute_match` as default value must be set by the onchange before the user selects the product. 
This will be fixed by #1390 so you can keep the code as it was.